src/libstd/collections/hash/raw/mod.rs

```diff
@@ -0,0 +1,1212 @@
+            RawIntoIter { iter, alloc }
+        }
+    }
+}
+
+/// Iterator over a a sub-range of a table. Unlike `RawIter` this iterator does
+/// not track an item count.
+pub struct RawIterRange<T> {
+    // Using *const here for covariance
+    data: *const T,
+    ctrl: *const u8,
+    current_group: BitMask,
+    end: *const u8,
+}
+
+impl<T> RawIterRange<T> {
+    /// Returns a `RawIterRange` covering a subset of a table.
+    ///
+    /// The start offset must be aligned to the group width.
+    #[inline]
+    unsafe fn new(
+        input_ctrl: *const u8,
+        input_data: *const T,
+        range: Range<usize>,
+    ) -> RawIterRange<T> {
+        debug_assert_eq!(range.start % Group::WIDTH, 0);
+        let ctrl = input_ctrl.add(range.start);
+        let data = input_data.add(range.start);
+        let end = input_ctrl.add(range.end);
+        debug_assert_eq!(offset_from(end, ctrl), range.end - range.start);
+        let current_group = Group::load_aligned(ctrl).match_empty_or_deleted().invert();
+        RawIterRange {
+            data,
+            ctrl,
+            current_group,
+            end,
+        }
+    }
+
+    /// Splits a `RawIterRange` into two halves.
+    ///
+    /// This will fail if the total range is smaller than the group width.
+    #[inline]
+    #[cfg(feature = "rayon")]
+    pub fn split(mut self) -> (RawIterRange<T>, Option<RawIterRange<T>>) {
+        unsafe {
+            let len = offset_from(self.end, self.ctrl);
+            debug_assert!(len.is_power_of_two());
+            if len <= Group::WIDTH {
+                (self, None)
+            } else {
+                debug_assert_eq!(len % (Group::WIDTH * 2), 0);
+                let mid = len / 2;
+                let tail = RawIterRange::new(self.ctrl, self.data, mid..len);
+                debug_assert_eq!(self.data.add(mid), tail.data);
+                debug_assert_eq!(self.end, tail.end);
+                self.end = self.ctrl.add(mid);
+                debug_assert_eq!(self.end, tail.ctrl);
+                (self, Some(tail))
+            }
+        }
+    }
+}
+
+unsafe impl<T> Send for RawIterRange<T> where T: Send {}
+unsafe impl<T> Sync for RawIterRange<T> where T: Sync {}
+
+impl<T> Clone for RawIterRange<T> {
+    #[inline]
+    fn clone(&self) -> Self {
+        RawIterRange {
+            data: self.data,
+            ctrl: self.ctrl,
+            current_group: self.current_group,
+            end: self.end,
+        }
+    }
+}
+
+impl<T> Iterator for RawIterRange<T> {
+    type Item = Bucket<T>;
+
+    #[inline]
+    fn next(&mut self) -> Option<Bucket<T>> {
+        unsafe {
+            loop {
+                if let Some(index) = self.current_group.lowest_set_bit() {
+                    self.current_group = self.current_group.remove_lowest_bit();
+                    return Some(Bucket::from_ptr(self.data.add(index)));
+                }
+
+                self.ctrl = self.ctrl.add(Group::WIDTH);
```

Comment:
NB as written this line is UB if you call next after this method returns None. (ptr::add must be in-bounds)

Also this makes the iterator really not a FusedIterator, because even if you make it wrapping_add, it may wrap and start returning random memory. (probably would just fault on the null page in most cases)

possible fix with minimal impact: 

```rust
if self.ctrl.offset_to(self.end) > Group::WIDTH as isize { 
   self.ctrl = self.ctrl.add(Group::WIDTH); 
} else { 
   return None; 
}
```